### DISCUSSION TITLE
Need example code of in-memory processing

### DISCUSSION THREAD

**Discussion by JohnHerry** [python-ffmpegio/python-ffmpegio#22] [Q&A] [answered]
Thanks for the good job.
    I have a processing task with ffmpeg like follows:

`
ffmpeg -i input.wav -ar 16000 -ac 1 -af loudnorm=I=-23 -acodec pcm_s16le  output.wav
`

in this process, it takes resample, rechannel, volume normalize and codec [from PCM24 to PCM16].

but now I want this process to be in memory, which means the input, or the output [or maybe both] is a numpy-array or some other in memory array data, instead of a file in disk, How to make such a process? thanks

**Comment by tikuma-lsuhsc** [comment, marked as answer]
@JohnHerry - (Moved your post from Issue to Discussions section.) You'd want to use `audio.filter()` function.

```python
import ffmpegio as ff
fsout , dataout= ff.audio.filter('loudnorm=I=-23', fsin, datain, ar=16000, ac=1)
```
Assuming you're using `ffmpegio` package and not just `ffmpegio-core`, here you prepare `datain` as a 2D Numpy array with column-wise channels and specify its sampling rate in `fsin`. The function returns the processed output data with its sampling rate first `fsout` followed by `dataout` numpy array.

**Reply by tikuma-lsuhsc** [reply]
One more. I left out "[from PCM24 to PCM16]" part. AFAIK, there is no way for you to specify 24-bit (3-byte) audio sample because integer data types are defined in the power-of-2 bits (`int8`, `int16`, `int32`, `int64`). So, if you're reading audio data from a PCM24 source, make sure that the samples are saved as `int32` and multiply by `2**8` to keep the data fidelity (i.e., converting from signed 1.15 fixed-point to 1.31 fixed-point notation by padding least significant zeros). Then to convert the sample format to signed 16 bit, add `sample_fmt='s16'` argument to the function call above.

**Reply by tikuma-lsuhsc** [reply]
One more tip. If you don't need to run any explicit filter, use `'anull'` filter and specify implicit filter options such as `ac` and `ar`

**Comment by JohnHerry** [comment]
Thank you very much

**Comment by tikuma-lsuhsc** [comment]
You're welcome. (I'm reopening this discussion as I want others to be able to see it easily)

**Comment by JohnHerry** [comment]
An  exception.
I use the sample code. the input audio is  sr=16000, ac=1, pcm16.
the target output audio is sr=16000, ac=1, pcm16, but the volume=-14
sample code:

```
from scipy.io import wavfile
import ffmpegio as ff

def audio_processor(input_file, target_fr, target_volume=-14):
    sr, audio_in = wavfile.read(input_file)
    fs_out, audio_out = ff.audio.filter(f"loudnorm=I={target_volume}", sr, audio_in, ar=target_fr, ac=1)
    audio_out = np.squeeze(audio_out)
    return audio_out


if __name__ == "__main__":
    audio_out = audio_processor("input.wav", 16000)
    wavfile.write("output.wav", 16000, audio_out)
```
The result audio file seems downsampled by the processing and left a mess.  that means when the input sample just fit part of the filter (in my example, samplling_rate and channels ),  the result will be wrong?

Python==3.9.9
ffmpegio==0.9.0
ffmpeg version 5.1

**Comment by tikuma-lsuhsc** [comment]
What do you mean by "the input sample just fit part of the filter "?

As far as I can tell on my end, `audio.filter` with your inputs work as intended. 

Please try to run it with debug logger on and check the executed ffmpeg command line.

My guess is that there is something funny going on with `dtype` of your input Numpy array.

Here is my test script:

```python
import ffmpegio as ff
import logging
import numpy as np

logging.basicConfig(level=logging.DEBUG)

x = np.random.randint(-2**15,2**15,(16000,2),np.int16)
fs = 16000

fs_out, y = ff.audio.filter(ff.filtergraph.loudnorm(I=-24), fs, x, ar=fs, ac=1)

print(fs_out)
print(y.shape)
```

This should print a line (among others):

```
DEBUG:ffmpegio:-nostdin -hide_banner -f s16le -c:a pcm_s16le -ac 2 -ar 16000 -sample_fmt s16 -i - 
                                          -ar 16000 -ac 1 -sample_fmt s16 -filter:a loudnorm=I=-24 -c:a pcm_s16le -f s16le -
```

If still not resolved, post your command line as well as dtype  of the input array.

P.S., BTW, you don't need to rely on `scipy.io` to read wav files and can also use `ffmpegio` (especially if you're not using `scipy` for other purpose)

```python
sr, audio_in = ff.audio.read(input_file)
```

**Reply by JohnHerry** [reply]
My input audio with only one channel. so
 -f s16le -c:a pcm_s16le -ac **1** -ar 16000 -sample_fmt s16 -i - 
                                          -ar 16000 -ac 1 -sample_fmt s16 -filter:a loudnorm=I=-24 -c:a pcm_s16le -f s16le -

**Reply by tikuma-lsuhsc** [reply]
Is this part of your `DEBUG:ffmpegio` log?

**Reply by JohnHerry** [reply]
Yes, this is from the debug log, I think the ffmpeg had view the input as two channel wavfile and so downsampled the audio to half, So my result audio seems lost half part inter frames.

**Reply by tikuma-lsuhsc** [reply]
Hmmm, if that is indeed what it run, that's something I've never seen before. So are you saying that if you input say 16000-sample int16 numpy array, you get 8000x2 int16 numpy array back?
What happens if you try

```python
fs_out, audio_out = ff.audio.filter("anull", sr, audio_in, ar=target_fr, ac=1)
```

or

```python
fs_out, audio_out = ff.audio.filter("anull", sr, audio_in)
```

**Comment by JohnHerry** [comment]
Here is an example
![1](https://github.com/python-ffmpegio/python-ffmpegio/assets/8011802/0e1cc735-1211-4bdf-aa57-e10d740188fd)

![2](https://github.com/python-ffmpegio/python-ffmpegio/assets/8011802/a2b0f0bc-5728-40df-9ade-0eae6dc4df14)

the result y.shape  is **(result_sample_length,  1)**,  not two-channel.
and the two "anull" examples are failed as before.

**Reply by tikuma-lsuhsc** [reply]
Could you check the shape of `audio_in`? Also, it would help you if you plot these waveforms.

I've run another test and it does what it supposed to do on my PC:

```python
import ffmpegio as ff
import logging
import numpy as np
from matplotlib import pyplot as plt

logging.basicConfig(level=logging.DEBUG)

sr = 16000
audio_in = (2**14 * np.sin(np.arange(sr) / sr * np.pi * 2 * 1000)).astype(np.int16)

ff_out, y = ff.audio.filter(
    ff.filtergraph.loudnorm(I=-24), sr, audio_in, ar=16000, ac=1
)

print(y.shape)
print(y.dtype)

plt.plot(audio_in, ".-")
plt.plot(y, ".-")
plt.legend(['audio_in','y'])
plt.show()
```

The output plot clearly shows 16000 samples in and out.

![Figure_1](https://github.com/python-ffmpegio/python-ffmpegio/assets/79113787/83f6cca8-bde8-465e-9ea2-6af63b276ccb)

**Reply by tikuma-lsuhsc** [reply]
Also, tested both FFmpeg 5.1 and 6.1. Both works the same way.
